Sources/Data Model/Audience/UserAttribute.swift

```diff
@@ -90,49 +90,55 @@ extension UserAttribute {
     
     func evaluate(attributes: OptimizelyAttributes?) throws -> Bool {
         
+        let conditionString = Utils.getConditionString(conditions: self)
+        
         // invalid type - parsed for forward compatibility only (but evaluation fails)
         if typeSupported == nil {
-            throw OptimizelyError.userAttributeInvalidType(self.type ?? "empty")
+            throw OptimizelyError.userAttributeInvalidType(conditionString)
         }
 
         // invalid match - parsed for forward compatibility only (but evaluation fails)
         guard let matchFinal = matchSupported else {
-            throw OptimizelyError.userAttributeInvalidMatch(self.match ?? "empty")
+            throw OptimizelyError.userAttributeInvalidMatch(conditionString)
         }
         
         guard let nameFinal = name else {
-            throw OptimizelyError.userAttributeInvalidFormat("empty name in condition")
+            throw OptimizelyError.userAttributeInvalidName(conditionString)
         }
         
         let attributes = attributes ?? OptimizelyAttributes()
         
-        let rawAttributeValue = attributes[nameFinal] ?? nil  // default to nil to avoid warning "coerced from 'Any??' to 'Any?'"
-        
+        let rawAttributeValue = attributes[nameFinal] ?? nil // default to nil to avoid warning "coerced from 'Any??' to 'Any?'"
+     
         if matchFinal != .exists {
+            if !attributes.keys.contains(nameFinal) {
+                throw OptimizelyError.missingAttributeValue(conditionString, nameFinal)
+            }
+
             if value == nil {
-                throw OptimizelyError.userAttributeInvalidFormat("missing value (\(nameFinal)) in condition)")
+                throw OptimizelyError.userAttributeNilValue(conditionString)
             }
             
             if rawAttributeValue == nil {
-                throw OptimizelyError.evaluateAttributeInvalidFormat("no attribute value for (\(nameFinal))")
+                throw OptimizelyError.nilAttributeValue(conditionString, nameFinal)
             }
         }
         
         switch matchFinal {
         case .exists:
             return !(rawAttributeValue is NSNull || rawAttributeValue == nil)
         case .exact:
-            return try value!.isExactMatch(with: rawAttributeValue!)
+            return try value!.isExactMatch(with: rawAttributeValue!, condition: conditionString, name: nameFinal)
```

Comment:
This case is tough. How can we avoid "conditionString" buiding overhead?

Reply:
Tried several ideas to evaluate `conditionString` just once for each `UserAttribute`. The problem i faced is that we are dealing with a struct which is a value type and to have a lazy or mutable property inside it, we would have to mark `evaluate` method as `mutating` which is not possible in our case since it is part of a protocol.

Reply:
One possible fix is to create "conditionsString" when UserAttribute is instantiated by datafile parsing (just like we did for "FeatureVariable.swift" for JSON type conversion. Then we can avoid "Utils.getConditionString".